flybywiresim/a32nx
# feat(efb): TOD calculator

## Summary of Changes

Main purpose: implement TOD (top of descend) calculator. 

On the occasion:
- Add new components (button, card, divider and update input component)
- Implement react redux

## Screenshots (if necessary)

**New screenshots in [comment](https://github.com/flybywiresim/a32nx/pull/3425#issuecomment-779520523)**

Old ones:

![2021-02-08 20_46_00-Microsoft Flight Simulator - 1 12 13 0](https://user-images.githubusercontent.com/13150438/107273601-903fe800-6a4f-11eb-8ef6-8562b1de8300.png)
![2021-02-08 20_46_43-Microsoft Flight Simulator - 1 12 13 0](https://user-images.githubusercontent.com/13150438/107273609-933ad880-6a4f-11eb-8dd7-2b5a0acc043c.png)
![2021-02-08 20_47_22-Microsoft Flight Simulator - 1 12 13 0](https://user-images.githubusercontent.com/13150438/107273640-9c2baa00-6a4f-11eb-89d0-9956c5c76222.png)
![2021-02-08 20_47_42-Microsoft Flight Simulator - 1 12 13 0](https://user-images.githubusercontent.com/13150438/107273649-9e8e0400-6a4f-11eb-9e07-796c72ec44ac.png)


## References


Manual calculations that led to the creation of class `src/Service/TODCalculator.tsx`.

Alghoritm:
- get GS layers provided by the user
- calculate average GS based on time user will spend in each of that layers
- knowing average GS we know how much it will take to reach specific distance - so we can calculate vertical speed
- knowing average GS and vertical speed we know how much it will take to reach specific altitude - so we can calculate distance needed

## Additional context



Discord username (if different from GitHub): arturkuma#6639

## Testing instructions

1. Check if *Tools* tool bar item is displayed
1. Make sure that calculations are correct
1. Ground speed works in ranges:

![2021-02-08 21_04_37-Microsoft Flight Simulator - 1 12 13 0](https://user-images.githubusercontent.com/13150438/107274837-3e985d00-6a51-11eb-873a-bf84fe143b28.png)

in that configuration GS is:
- 250 KN between 0 and 10 000 FT
- 360 KN between 10 000 and 20 000 FT
- 436 KN above 20 000

order of GS layers is not important - alghoritm will sort it out


## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link
